src/gnu.rs

```diff
@@ -0,0 +1,109 @@
+use std::borrow::Cow;
+use std::cmp;
+use std::io::prelude::*;
+use std::io::{self, SeekFrom};
+use std::path::Path;
+
+use header::bytes2path;
+use other;
+use {Entry, Header};
+
+macro_rules! try_iter {
+    ($e:expr) => (match $e {
+        Ok(e) => e,
+        Err(e) => return Some(Err(e)),
+    })
+}
+
+/// dox
+pub struct GnuEntries<'a, R: 'a> {
+    inner: Box<Iterator<Item=io::Result<Entry<'a, R>>> + 'a>,
+}
+
+/// dox
+pub struct GnuEntry<'a, R: 'a> {
+    inner: Entry<'a, R>,
+    name: Option<Vec<u8>>,
+}
+
+impl<'a, R: 'a + Read> GnuEntries<'a, R> {
+    /// dox
+    pub fn new<I>(i: I) -> GnuEntries<'a, R>
+        where I: IntoIterator<Item=io::Result<Entry<'a, R>>> + 'a,
+              I::IntoIter: 'a,
+    {
+        GnuEntries { inner: Box::new(i.into_iter()) }
+    }
+}
+
+impl<'a, R: 'a + Read> Iterator for GnuEntries<'a, R> {
+    type Item = io::Result<GnuEntry<'a, R>>;
+
+    fn next(&mut self) -> Option<io::Result<GnuEntry<'a, R>>> {
+        let mut entry = match self.inner.next() {
+            Some(Ok(e)) => e,
+            Some(Err(e)) => return Some(Err(e)),
+            None => return None,
+        };
+
+        if !entry.header().entry_type().is_gnu_longname() {
+            return Some(Ok(GnuEntry { inner: entry, name: None }))
+        }
+
+        // Don't allow too too crazy allocation sizes up front
+        let cap = cmp::min(entry.header().size().unwrap_or(0), 128 * 1024);
+        let mut filename = Vec::with_capacity(cap as usize);
+        try_iter!(entry.read_to_end(&mut filename));
```

Comment:
Yeah it's definitely possible that this could go off the rails, which is a concern of @aidanhs's I believe for handling things such as malicious archives.